util/pkg/vfs/ossfs.go

```diff
@@ -0,0 +1,358 @@
+// we take a process-wide lock during the operation.
+// Not a great approach, but fine for a single process (with low concurrency)
+// TODO: should we enable versioning?
+var createFileLockOSS sync.Mutex
+
+func (p *OSSPath) CreateFile(data io.ReadSeeker, acl ACL) error {
+	createFileLockOSS.Lock()
+	defer createFileLockOSS.Unlock()
+
+	// Check if exists
+	b := p.client.Bucket(p.bucket)
+	exist, err := b.Exists(p.key)
+	if err != nil {
+		return err
+	}
+	if exist {
+		return os.ErrExist
+	}
+
+	return p.WriteFile(data, acl)
+}
+
+func (p *OSSPath) Remove() error {
+	b := p.client.Bucket(p.bucket)
+
+	done, err := RetryWithBackoff(ossWriteBackoff, func() (bool, error) {
+		glog.V(8).Infof("removing file %s", p)
+
+		err := b.Del(p.key)
+		if err != nil {
+			return false, fmt.Errorf("error deleting %s: %v", p, err)
+		}
+		return true, nil
+	})
+	if err != nil {
+		return err
+	} else if done {
+		return nil
+	} else {
+		// Shouldn't happen - we always return a non-nil error with false
+		return wait.ErrWaitTimeout
+	}
+}
+
+func (p *OSSPath) Base() string {
+	return path.Base(p.key)
+}
+
+func (p *OSSPath) Path() string {
+	return "oss://" + p.bucket + "/" + p.key
+}
+
+func (p *OSSPath) ReadDir() ([]Path, error) {
+	prefix := p.key
+	if !strings.HasSuffix(prefix, "/") {
+		prefix += "/"
+	}
+	// OSS can return at most 1000 paths(keys + common prefixes) at a time
+	opt := listOption{
+		prefix: prefix,
+		delim:  "/",
+		marker: "",
+		max:    1000,
+	}
+
+	return p.listPath(opt)
+}
+
+func (p *OSSPath) ReadTree() ([]Path, error) {
+	prefix := p.key
+	if !strings.HasSuffix(prefix, "/") {
+		prefix += "/"
+	}
+	// OSS can return at most 1000 paths(keys + common prefixes) at a time
+	opt := listOption{
+		prefix: prefix,
+		// No delimiter for recursive search
+		delim:  "",
+		marker: "",
+		max:    1000,
+	}
+
+	return p.listPath(opt)
+}
+
+func (p *OSSPath) PreferredHash() (*hashing.Hash, error) {
+	return p.Hash(hashing.HashAlgorithmMD5)
+}
+
+func (p *OSSPath) Hash(a hashing.HashAlgorithm) (*hashing.Hash, error) {
+	if a != hashing.HashAlgorithmMD5 {
+		return nil, nil
+	}
+
+	md5 := p.hash
+	if md5 == "" {
+		return nil, nil
+	}
+
+	md5Bytes, err := hex.DecodeString(md5)
+	if err != nil {
+		return nil, fmt.Errorf("Etag was not a valid MD5 sum: %q", md5)
+	}
+
+	return &hashing.Hash{Algorithm: hashing.HashAlgorithmMD5, HashValue: md5Bytes}, nil
+}
+
+func (p *OSSPath) listPath(opt listOption) ([]Path, error) {
+	var ret []Path
+	b := p.client.Bucket(p.bucket)
+
+	done, err := RetryWithBackoff(ossReadBackoff, func() (bool, error) {
+
+		var paths []Path
+		for {
+			// OSS can return at most 1000 paths(keys + common prefixes) at a time
+			resp, err := b.List(opt.prefix, opt.delim, opt.marker, opt.max)
+			if err != nil {
+				if isOSSNotFound(err) {
+					return true, os.ErrNotExist
+				}
+				return false, fmt.Errorf("error listing %s: %v", p, err)
+			}
+
+			conLen := len(resp.Contents)
+			comPrefLen := len(resp.CommonPrefixes)
```

Comment:
suggestion: I'd probably not declare the variables, and just use `len(resp.Contents)` (for example) directly where we have conLen.  Perf should be unnoticeably similar, and it makes it a little more readable IMO

Reply:
Agree.